module/zfs/zio.c

```diff
@@ -890,36 +890,69 @@ zio_root(spa_t *spa, zio_done_func_t *done, void *private, enum zio_flag flags)
 	return (zio_null(NULL, spa, NULL, done, private, flags));
 }
 
-static void
-zfs_blkptr_verify(spa_t *spa, const blkptr_t *bp, boolean_t config_held)
+/*
+ * Allow two modes for zfs_blkptr_verify:
+ *    1) BLK_VERIFY_ONLY (only evalulate the block pointer)
+ *    2) BLK_VERIFY_HALT (call zfs_panic_remover on error)
+ */
+static boolean_t
+zio_panic_recover(enum blk_verify_flag blk_verify, const char *fmt, ...)
+{
+	/* determine if the caller should return (B_TRUE) or keep checking */
+	if (blk_verify == BLK_VERIFY_ONLY)
+		return (B_TRUE);
+
+	zfs_panic_recover(fmt);
```

Comment:
I implemented this for checking all the blkptr_t fields.  Down in the DVA validation, there are continue statements after the zfs_panic_recover calls.  So I probably can't just return a boolean result of zfs_blkptr_verify_impl there.  Should I be checking zfs_recover to decide whether to continue?

Reply:
I see.  Then perhaps the cleanest thing to do would be to set a `valid` flag for all of the checks and return that.

This way when `BLK_VERIFY_ONLY` is set all errors in a block pointer are logged (instead of just the first).  When `BLK_VERIFY_HALT` is set the behavior is unchanged aside from the new return value.

I'm thinking `zfs_blkptr_verify_log` would be a better name for the wrapper function as well.  What do you think about this:

```
static void
zfs_blkptr_verify(spa_t *spa, const blkptr_t *bp, boolean_t config_held)
{
        boolean_t valid = B_TRUE; 

        if (!DMU_OT_IS_VALID(BP_GET_TYPE(bp))) {
                valid = zfs_blkptr_verify_log(
                    "blkptr at %p has invalid TYPE %llu",
                    bp, (longlong_t)BP_GET_TYPE(bp));
        }
        ...

        for (int i = 0; i < BP_GET_NDVAS(bp); i++) {
                uint64_t vdevid = DVA_GET_VDEV(&bp->blk_dva[i]);

                if (vdevid >= spa->spa_root_vdev->vdev_children) {
                        valid = zfs_blkptr_verify_log(
                            "blkptr at %p DVA %u has invalid VDEV %llu",
                            bp, i, (longlong_t)vdevid);
                        continue;
                }
                ...
        }
        ...

        return (valid);
}
```

Reply:
Thanks I'll give this a try.

Reply:
Here's a sample.
{code}
# zdb -G -R tank 0:9001c400:400:b
Found vdev: /root/file1
DVA[0]=<0:6041e00:200> [L0 ZFS plain file] fletcher4 lz4 unencrypted LE contiguous unique single size=400L/200P birth=623821L/623821P fill=1 cksum=239a072c46:9690c9e7f2d:1a8b7b42bdbb0:383da26fd31df3
ZFS_DBGMSG(zdb) START:
spa.c:5051:spa_open_common(): spa_open_common: opening tank
spa_misc.c:410:spa_load_note(): spa_load(tank, config trusted): LOADING
vdev.c:125:vdev_dbgmsg(): file vdev '/root/file1': best uberblock found for spa tank. txg 1482823
spa_misc.c:410:spa_load_note(): spa_load(tank, config untrusted): using uberblock with txg=1482823
spa_misc.c:410:spa_load_note(): spa_load(tank, config trusted): spa_load_verify found 0 metadata errors and 1 data errors
spa_misc.c:410:spa_load_note(): spa_load(tank, config trusted): LOADED
zio.c:914:zfs_blkptr_verify_log(): tank: blkptr at 0x3482000 has invalid TYPE 158
zio.c:914:zfs_blkptr_verify_log(): tank: blkptr at 0x3482000 has invalid CHECKSUM 127
zio.c:914:zfs_blkptr_verify_log(): tank: blkptr at 0x3482000 has invalid COMPRESS 45
zio.c:914:zfs_blkptr_verify_log(): tank: blkptr at 0x3482000 DVA 0 has invalid VDEV 290
zio.c:914:zfs_blkptr_verify_log(): tank: blkptr at 0x3482000 DVA 1 has invalid VDEV 5177392
zio.c:914:zfs_blkptr_verify_log(): tank: blkptr at 0x3482000 DVA 2 has invalid VDEV 10240
ZFS_DBGMSG(zdb) END
{\code}

Reply:
Was talking with Tom about this and he thinks we shouldn't be printing out all these ominous messages for something that ultimately worked on the retry.  I could further refactor this to defer the printing either to the caller or to a different interface (or another BLK_VERIFY enum) that gets called when zfs_blkptr_verify is FALSE... what's your thinking on this?

Reply:
Adding another enum which could be used by `zdb` seems like a reasonable way to go.  Perhaps, `BLK_VERIFY_LOG` (which does log), `BLK_VERIFY_ONLY` (which does not log), and `BLK_VERIFY_HALT`.  Then the caller would be able to determine if it merits logging.

Offhand, the only place we might not want to log this would be `zdb`.  But even there you'd need to specify `-G` to see the errors.  In all other places, the block pointer should always be valid.  If it's ever not that's noteworthy and I think worth logging.